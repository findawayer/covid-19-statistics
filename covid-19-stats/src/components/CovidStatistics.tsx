import type { FunctionComponent } from 'react';
import React from 'react';

// import Dump from '../components/Dump';
import { useFetch } from '../hooks/useFetch';
import { useTabularData } from '../hooks/useTabularData';
import DataTable from './DataTable';
import Search from './Search';
import Loading from './Loading';
import {
  StyledCovidStatHeader,
  StyledCovidStatTitle
} from './styles/CovidStatistics';

const API_ALL_COUNTRY = 'https://covid19-api.com/country/all';

type CountryData = {
  country: string;
  code: string;
  confirmed: number;
  recovered: number;
  critical: number;
  deaths: number;
  latitude: number;
  longitude: number;
  lastChange: string; // Date string
  lastUpdate: string; // Date string
};

const columns = [
  { key: 'country', alias: '국가' },
  { key: 'confirmed', alias: '확진자' },
  { key: 'recovered', alias: '완치자' },
  { key: 'critical', alias: '치명적' },
  { key: 'deaths', alias: '사망자' }
];

interface CovidStatTableProps {}

const CovidStatTable: FunctionComponent<CovidStatTableProps> = () => {
  const { data, loading, error } = useFetch<CountryData[]>(API_ALL_COUNTRY);
  const {
    processedData,
    handleColumnClick,
    handleKeywordChange
  } = useTabularData(data, {
    filterKey: 'country',
    sortKey: 'country',
    sortDirection: 'ascending'
  });

  if (loading) return <Loading />;
  if (!data) return <p>No data!</p>;
  if (error) return <p>{error.message}</p>;

  return (
    <>
      <StyledCovidStatHeader>
        <StyledCovidStatTitle>코로나 바이러스 세계 현황</StyledCovidStatTitle>
        <Search handleChange={handleKeywordChange} />
      </StyledCovidStatHeader>
      <DataTable
        data={processedData}
        columns={columns}
        handleColumnClick={handleColumnClick}
      />
    </>
  );
};

export default CovidStatTable;
