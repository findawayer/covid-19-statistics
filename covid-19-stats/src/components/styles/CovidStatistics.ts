import styled from '@emotion/styled';

export const StyledCovidStatHeader = styled.div`
  display: flex;
  padding: 10px 0;
  align-items: center;
  justify-content: space-between;
`;

export const StyledCovidStatTitle = styled.h1`
  font-size: 2rem;
  font-weight: 700;
`;
