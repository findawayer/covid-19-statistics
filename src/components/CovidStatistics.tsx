import type { FunctionComponent } from 'react';
import React from 'react';

import { API_COVID_19_ALL_COUNTRY } from '../config/api';
import { useFetch } from '../hooks/useFetch';
import { useTabularData } from '../hooks/useTabularData';
import type { CountryData } from '../models/covid-19-api';
import DataTable from './DataTable';
import Loading from './Loading';
import Search from './Search';
import {
  StyledCovidStatBody,
  StyledCovidStatHeader,
  StyledCovidStatScroll,
  StyledCovidStatTitle,
} from './styles/CovidStatistics';

const columns = [
  { key: 'country', alias: '국가' },
  { key: 'confirmed', alias: '확진자' },
  { key: 'recovered', alias: '완치자' },
  { key: 'critical', alias: '치명적' },
  { key: 'deaths', alias: '사망자' },
];

const CovidStatTable: FunctionComponent = () => {
  const { data, loading, error } = useFetch<CountryData[]>(
    API_COVID_19_ALL_COUNTRY,
    { cacheKey: 'covid19-country-all', cacheMaxAge: 60 * 60 * 1000 }, // 1 hour cache
  );
  const {
    processedData,
    dataOptions: { sortKey, sortDirection },
    handleColumnClick,
    handleKeywordChange,
  } = useTabularData(data, {
    filterKey: 'country',
    sortKey: 'country',
    sortDirection: 'ascending',
  });

  const renderTable = () => {
    if (loading) return <Loading />;
    if (error) return <p>{error.message}</p>;
    if (!data) return <p>No data!</p>;
    return (
      <DataTable
        data={processedData}
        columns={columns}
        sortKey={sortKey}
        sortDirection={sortDirection}
        handleColumnClick={handleColumnClick}
      />
    );
  };

  return (
    <>
      <StyledCovidStatHeader>
        <StyledCovidStatTitle>코로나 바이러스 세계 현황</StyledCovidStatTitle>
        <Search handleChange={handleKeywordChange} />
      </StyledCovidStatHeader>
      <StyledCovidStatBody>
        <StyledCovidStatScroll>{renderTable()}</StyledCovidStatScroll>
      </StyledCovidStatBody>
    </>
  );
};

export default CovidStatTable;
