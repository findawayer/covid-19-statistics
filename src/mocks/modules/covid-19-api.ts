import type { CovidCountryStat } from '../../models/covid-19-api';

export const covidCountryStatMock: CovidCountryStat[] = [
  {
    code: 'KR',
    confirmed: 68664,
    country: 'South Korea',
    critical: 401,
    deaths: 1125,
    lastChange: '2021-01-10T06:16:39+01:00',
    lastUpdate: '2021-01-10T09:30:04+01:00',
    latitude: 35.907757,
    longitude: 127.766922,
    recovered: 50409,
  },
  {
    code: 'MN',
    confirmed: 1429,
    country: 'Mongolia',
    critical: 14,
    deaths: 2,
    lastChange: '2021-01-10T06:24:51+01:00',
    lastUpdate: '2021-01-10T09:30:04+01:00',
    latitude: 46.862496,
    longitude: 103.846656,
    recovered: 896,
  },
  {
    code: 'TW',
    confirmed: 828,
    country: 'Taiwan',
    critical: 0,
    deaths: 7,
    lastChange: '2021-01-09T07:15:52+01:00',
    lastUpdate: '2021-01-10T09:30:04+01:00',
    latitude: 23.69781,
    longitude: 120.960515,
    recovered: 720,
  },
];
