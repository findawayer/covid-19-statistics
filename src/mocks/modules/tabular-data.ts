export const tabularData = [
  { name: 'foo', size: 10 },
  { name: 'bar', size: 500 },
  { name: 'baz', size: 1000 },
  { name: 'qux', size: 1000000 },
];

export const tableColumns = [
  { key: 'name', alias: '이름' },
  { key: 'size', alias: '크기' },
];
