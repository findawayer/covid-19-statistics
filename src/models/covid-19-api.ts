export type CountryData = {
  country: string;
  code: string;
  confirmed: number;
  recovered: number;
  critical: number;
  deaths: number;
  latitude: number;
  longitude: number;
  lastChange: string; // Date string
  lastUpdate: string; // Date string
};
