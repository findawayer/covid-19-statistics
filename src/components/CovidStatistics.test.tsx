import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { covidCountryStatMock } from '../mocks/modules/covid-19-api';
import CovidStatistics from './CovidStatistics';

describe('<CovidStatistics />', () => {
  const useFetch = jest.spyOn(require('../hooks/useFetch'), 'useFetch');

  describe('without any data', () => {
    afterEach(() => {
      useFetch.mockClear();
    });

    it('renders a spinner while loading data', async () => {
      useFetch.mockReturnValue({
        data: null,
        error: null,
        loading: true,
      });
      const { container } = render(<CovidStatistics />);

      expect(useFetch).toBeCalledTimes(1);
      expect(container.querySelector('svg')).toBeInTheDocument();
    });

    it('renders an error message with no data fetched', async () => {
      useFetch.mockReturnValue({
        data: null,
        error: new Error('Request failed.'),
        loading: false,
      });
      const { getByText } = render(<CovidStatistics />);

      expect(useFetch).toBeCalledTimes(1);
      expect(getByText('Request failed.')).toBeInTheDocument();
    });
  });

  describe('with data', () => {
    beforeEach(() => {
      useFetch.mockReturnValue({
        data: covidCountryStatMock,
        error: null,
        loading: false,
      });
    });

    it('renders COVID stat table with data fetched', async () => {
      const { getByText } = render(<CovidStatistics />);

      expect(useFetch).toBeCalledTimes(1);
      expect(getByText(/South Korea/)).toBeInTheDocument();
    });

    it('filters data when search field receives a value (South Korea)', async () => {
      const { getByPlaceholderText, getByTestId, queryByText } = render(
        <CovidStatistics />,
      );
      // Enter search keyword 'K'
      const searchField = getByPlaceholderText('국가로 검색');
      userEvent.type(searchField, 'K');
      // Mongolia is filtered out
      await waitFor(() => expect(queryByText('Mongolia')).toBeNull());
      expect(getByTestId('tbody')).toMatchSnapshot();
    });

    it('sorts data upside down when the active column heading is clicked (Taiwan, South Korea, Mongolia)', async () => {
      const { getByTestId } = render(<CovidStatistics />);
      // Click country
      userEvent.click(getByTestId('country'));
      expect(getByTestId('tbody')).toMatchSnapshot();
    });

    it('reverts data order back to original when same column is clicked twice (Mongolia, South Korea, Taiwan)', async () => {
      const { getByTestId } = render(<CovidStatistics />);
      // Double click country
      userEvent.dblClick(getByTestId('country'));
      expect(getByTestId('tbody')).toMatchSnapshot();
    });

    it('sorts data in descending direction when a column heading is clicked (South Korea, Monglia, Taiwan)', async () => {
      const { getByTestId } = render(<CovidStatistics />);
      // Click country
      userEvent.click(getByTestId('confirmed'));
      expect(getByTestId('tbody')).toMatchSnapshot();
    });
  });
});
