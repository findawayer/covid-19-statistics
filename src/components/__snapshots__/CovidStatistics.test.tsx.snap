// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CovidStatistics /> with data filters data when search field receives a value (South Korea) 1`] = `
<tbody
  data-testid="tbody"
>
  <tr>
    <td
      class="css-1g8ku9a"
    >
      South Korea
    </td>
    <td
      class="css-1g8ku9a"
    >
      68,664
    </td>
    <td
      class="css-1g8ku9a"
    >
      50,409
    </td>
    <td
      class="css-1g8ku9a"
    >
      401
    </td>
    <td
      class="css-1g8ku9a"
    >
      1,125
    </td>
  </tr>
</tbody>
`;

exports[`<CovidStatistics /> with data reverts data order back to original when same column is clicked twice (Mongolia, South Korea, Taiwan) 1`] = `
<tbody
  data-testid="tbody"
>
  <tr>
    <td
      class="css-1g8ku9a"
    >
      Mongolia
    </td>
    <td
      class="css-1g8ku9a"
    >
      1,429
    </td>
    <td
      class="css-1g8ku9a"
    >
      896
    </td>
    <td
      class="css-1g8ku9a"
    >
      14
    </td>
    <td
      class="css-1g8ku9a"
    >
      2
    </td>
  </tr>
  <tr>
    <td
      class="css-1g8ku9a"
    >
      South Korea
    </td>
    <td
      class="css-1g8ku9a"
    >
      68,664
    </td>
    <td
      class="css-1g8ku9a"
    >
      50,409
    </td>
    <td
      class="css-1g8ku9a"
    >
      401
    </td>
    <td
      class="css-1g8ku9a"
    >
      1,125
    </td>
  </tr>
  <tr>
    <td
      class="css-1g8ku9a"
    >
      Taiwan
    </td>
    <td
      class="css-1g8ku9a"
    >
      828
    </td>
    <td
      class="css-1g8ku9a"
    >
      720
    </td>
    <td
      class="css-1g8ku9a"
    >
      0
    </td>
    <td
      class="css-1g8ku9a"
    >
      7
    </td>
  </tr>
</tbody>
`;

exports[`<CovidStatistics /> with data sorts data in descending direction when a column heading is clicked (South Korea, Monglia, Taiwan) 1`] = `
<tbody
  data-testid="tbody"
>
  <tr>
    <td
      class="css-1g8ku9a"
    >
      South Korea
    </td>
    <td
      class="css-1g8ku9a"
    >
      68,664
    </td>
    <td
      class="css-1g8ku9a"
    >
      50,409
    </td>
    <td
      class="css-1g8ku9a"
    >
      401
    </td>
    <td
      class="css-1g8ku9a"
    >
      1,125
    </td>
  </tr>
  <tr>
    <td
      class="css-1g8ku9a"
    >
      Mongolia
    </td>
    <td
      class="css-1g8ku9a"
    >
      1,429
    </td>
    <td
      class="css-1g8ku9a"
    >
      896
    </td>
    <td
      class="css-1g8ku9a"
    >
      14
    </td>
    <td
      class="css-1g8ku9a"
    >
      2
    </td>
  </tr>
  <tr>
    <td
      class="css-1g8ku9a"
    >
      Taiwan
    </td>
    <td
      class="css-1g8ku9a"
    >
      828
    </td>
    <td
      class="css-1g8ku9a"
    >
      720
    </td>
    <td
      class="css-1g8ku9a"
    >
      0
    </td>
    <td
      class="css-1g8ku9a"
    >
      7
    </td>
  </tr>
</tbody>
`;

exports[`<CovidStatistics /> with data sorts data upside down when the active column heading is clicked (Taiwan, South Korea, Mongolia) 1`] = `
<tbody
  data-testid="tbody"
>
  <tr>
    <td
      class="css-1g8ku9a"
    >
      Taiwan
    </td>
    <td
      class="css-1g8ku9a"
    >
      828
    </td>
    <td
      class="css-1g8ku9a"
    >
      720
    </td>
    <td
      class="css-1g8ku9a"
    >
      0
    </td>
    <td
      class="css-1g8ku9a"
    >
      7
    </td>
  </tr>
  <tr>
    <td
      class="css-1g8ku9a"
    >
      South Korea
    </td>
    <td
      class="css-1g8ku9a"
    >
      68,664
    </td>
    <td
      class="css-1g8ku9a"
    >
      50,409
    </td>
    <td
      class="css-1g8ku9a"
    >
      401
    </td>
    <td
      class="css-1g8ku9a"
    >
      1,125
    </td>
  </tr>
  <tr>
    <td
      class="css-1g8ku9a"
    >
      Mongolia
    </td>
    <td
      class="css-1g8ku9a"
    >
      1,429
    </td>
    <td
      class="css-1g8ku9a"
    >
      896
    </td>
    <td
      class="css-1g8ku9a"
    >
      14
    </td>
    <td
      class="css-1g8ku9a"
    >
      2
    </td>
  </tr>
</tbody>
`;
